### DISCUSSION TITLE
Parallel igl::harmonic

### DISCUSSION THREAD

**Discussion by HollowSun0** [libigl/libigl#1910] [Ideas]
Hi @alecjacobson,

Thank you so much for all your efforts put into this wonderful and revolutionary library! I have one question regarding [this igl::harmonic method](https://github.com/libigl/libigl/blob/36e8435a2f724e83e14f79d128102d06b514a4f4/include/igl/harmonic.cpp#L26). When passing enormous matrices to this method, calculations can take a while; sometimes about 10 hours for k = 3 and 25 gigs of RAM. Is there any way to parallelize this operation / make it faster? RAM trade-offs are acceptable.

We are currently testing Eigen with OpenMP but I'm not sure how efficient it will be.

Thank you,
Patrick Laurin

**Comment by alecjacobson** [comment]
I'm guessing that the bottleneck will be the solve (factorization and back substitution) and not the construction of the matrices.

If that's the case, then you'd need to switch from `min_quad_with_fixed` which uses the Eigen cholesky solver to a parallel solver like MKL or SuiteSparse. These are rather easy to invoke from Eigen (albeit not officially supported). You might have a look here https://github.com/alecjacobson/sparse-solver-benchmark to get a sense of how to call them.

**Reply by HollowSun0** [reply]
I've tried the SuiteSparse's `Eigen::CholmodSupernodalLLT<Eigen::SparseMatrix<T > >` solver but it seems like simply changing the solver is not enough. In cases where the matrix is not positive definite, CHOLMOD will emit a warning and result in a numerical issue: `CHOLMOD warning: not positive definite. file: C:\dev\git\suitesparse\SuiteSparse\CHOLMOD\Cholesky\t_cholmod_rowfac.c line: 430`. I'm not sure if a fallback to LDLT will make it work.

**Reply by HollowSun0** [reply]
From what I understand, igl::harmonic only accepts positive definite matrices. I must've had the same error before changing the solver.

**Reply by alecjacobson** [reply]
The matrix in igl::harmonic really should be positive definite. It should be semi-positive before constraints and positive definite once constraints are factored out.

**Reply by HollowSun0** [reply]
Works for me. Thanks!
